### PR TITLE
[DC-2531] add clientName support

### PR DESCRIPTION
**Pull request by lxol**

### CODE REVIEW

**Review by elabeca — APPROVED** [review]

File: app/uk/gov/hmrc/contactadvisors/controllers/EmailController.scala [hunk 1]
```diff
@@ -50,7 +50,9 @@ class EmailController @Inject()(
     {
       implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers)
       authorised(predicates.foldLeft[Predicate](AuthProviders(PrivilegedApplication))((b, a) => b and a)) {
-        request.body.asJson.fold(Future.successful(BadRequest("""{"error": "invalid payload"}""")))((json: JsValue) => emailService.doSendEmail(json))
+        request.body.asJson
+          .flatMap(_.asOpt[JsObject])
+          .fold(Future.successful(BadRequest("""{"error": "invalid payload"}""")))((jsobj: JsObject) => emailService.doSendEmail(jsobj))
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/contactadvisors/controllers/EmailController.scala:55]: do we want to return a more specific error of what is wrong? 
Also maybe log the exception as kibana error log message?

File: app/uk/gov/hmrc/contactadvisors/service/EmailService.scala [hunk 2]
```diff
@@ -29,17 +29,15 @@ import scala.concurrent.{ ExecutionContext, Future }
...
-    .getOptional[Seq[String]](s"email.clientIds")
-    .getOrElse(List())
-    .toSet
+    .getOptional[Map[String, String]](s"email.clientId")
+    .getOrElse(Map())
 
-  def doSendEmail(payload: JsValue)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Result] =
+  def doSendEmail(payload: JsObject)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Result] =
     if (!validParameters(payload)) {
       Future.successful(BadRequest("""{"error": "invalid parameters"}"""))
```
Comment (OlegEfrem) [inline app/uk/gov/hmrc/contactadvisors/service/EmailService.scala:37]: same thing here, do we want a more specific error to tell what parameters are wrong?

**Review by rwalpole — APPROVED** [review]

File: app/uk/gov/hmrc/contactadvisors/service/EmailService.scala [hunk 3]
```diff
@@ -49,11 +47,9 @@ class EmailService @Inject()(emailConnector: EmailConnector, conf: Configuration
...
 
-  private def validClientId(payload: JsValue): Boolean =
-    (payload \ "clientId").toOption
-      .collect {
-        case x: JsString if clientIds.contains(x.value) => true
-      }
-      .getOrElse(false)
+  private def getClientName(payload: JsValue): Option[String] =
+    (payload \ "clientId").toOption.collect {
+      case x: JsString => clientIds.collectFirst { case (key, value) if value == x.value => key }
```
Comment (muralishub) [inline app/uk/gov/hmrc/contactadvisors/service/EmailService.scala:52]: clientIds no more a collection is int it ?

File: test/uk/gov/hmrc/contactadvisors/service/EmailServiceSpec.scala [hunk 4]
```diff
@@ -166,6 +167,6 @@ class EmailServiceSpec extends PlaySpec with MockitoSugar with ScalaFutures with
                                              |        "name": "Mr Joe Bloggs"
                                              |  }
                                              |}
-        """.stripMargin)
+        """.stripMargin).as[JsObject]
 
 }
```
Comment (OlegEfrem) [inline test/uk/gov/hmrc/contactadvisors/service/EmailServiceSpec.scala:172]: do we have the newly adder error cases covered in a unit test?

File: app/uk/gov/hmrc/contactadvisors/service/EmailService.scala (hunk 3 above)
Comment (lxol) [reply app/uk/gov/hmrc/contactadvisors/service/EmailService.scala:52]: it is a collection, it is a Map

File: test/uk/gov/hmrc/contactadvisors/service/EmailServiceSpec.scala (hunk 4 above)
Comment (lxol) [reply test/uk/gov/hmrc/contactadvisors/service/EmailServiceSpec.scala:172]: there is no newly added error

File: app/uk/gov/hmrc/contactadvisors/controllers/EmailController.scala (hunk 1 above)
Comment (lxol) [reply app/uk/gov/hmrc/contactadvisors/controllers/EmailController.scala:55]: it is not an error of the microservice but the client error. If a careless client sends a lot of invalid request we will receive PagerDuty alerts.

**Review by OlegEfrem — APPROVED** [review]

**Pull request merged**